pkg/services/secrets/kvstore/sql.go

```diff
@@ -0,0 +1,151 @@
+package kvstore
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/secrets"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+// secretsKVStoreSQL provides a key/value store backed by the Grafana database
+type secretsKVStoreSQL struct {
+	log             log.Logger
+	sqlStore        sqlstore.Store
+	secretsService  secrets.Service
+	decryptionCache decryptionCache
+}
+
+type decryptionCache struct {
+	cache map[int64]cachedDecrypted
+	sync.Mutex
+}
+
+type cachedDecrypted struct {
+	updated time.Time
+	value   string
+}
+
+// Get an item from the store
+func (kv *secretsKVStoreSQL) Get(ctx context.Context, orgId int64, namespace string, typ string) (string, bool, error) {
+	item := Item{
+		OrgId:     &orgId,
+		Namespace: &namespace,
+		Type:      &typ,
+	}
+	var itemFound bool
+	var decryptedValue []byte
+
+	err := kv.sqlStore.WithDbSession(ctx, func(dbSession *sqlstore.DBSession) error {
+		has, err := dbSession.Get(&item)
+		if err != nil {
+			kv.log.Debug("error getting secret value", "orgId", orgId, "type", typ, "namespace", namespace, "err", err)
+			return err
+		}
+		if !has {
+			kv.log.Debug("secret value not found", "orgId", orgId, "type", typ, "namespace", namespace)
+			return nil
+		}
+		itemFound = true
+		kv.log.Debug("got secret value", "orgId", orgId, "type", typ, "namespace", namespace, "value", item.Value)
+		return nil
+	})
+
+	if err == nil && itemFound {
+		kv.decryptionCache.Lock()
+		defer kv.decryptionCache.Unlock()
+
+		if cache, present := kv.decryptionCache.cache[item.Id]; present && item.Updated.Equal(cache.updated) {
+			return cache.value, itemFound, err
+		}
+
+		decryptedValue, err = kv.secretsService.Decrypt(ctx, []byte(item.Value))
+		kv.decryptionCache.cache[item.Id] = cachedDecrypted{
+			updated: item.Updated,
+			value:   string(decryptedValue),
+		}
+	}
+
+	return string(decryptedValue), itemFound, err
```

Comment:
nit: I use to prefer to make explicit returns to make it clear if we expect a failure return or a successful one. This last one seems to be a failure, so I would prefer something like `nil, false, err` and for the successful cases `string(decryptedValue), true, nil`

Reply:
You are right, in the case of string should be "" and is not a failure return, I didn't see there is no explicit return after the call to `kv.secretsService.Decrypt`. Maybe this was not the case but if we know its empty and we are returning an error I was thinking about returning the value when we know which the value would be for example, if we know that at this point `err` should be `nil`, return that `nil` instead of the name of the variable `err`. For example, inside the first `if err == nil` we are returning ` err` when we know it is `nil`. But is only for readability.